### ISSUE TITLE
Teaching assistant responsibilities

### ISSUE THREAD

**Issue by SaraMati** [BME1478H/BME1478material#11]
Responsibilities are initially divided up as outlined below. As this is the first time we are running this course, these responsibilities may change as the course progresses.

Wednesday class:

- (Sara, Lina, Ryan, Maryam) TAs attending in RM 412, Rosebrugh building from 10 AM – 12 PM. Depending on student attendance, can also be answering online enquiries. Via Slack, email?

- (Joel) handling online enquiries. Via Slack, email?

Course development: (Sara, Lina and Joel)

Marking assignments:

- (Dae and Andrew). Initially responsible for marking weekly exercises and assignments only. This may change if we experience excessive demand at Wednesday sessions.

- (Maryam, Ryan). Partly responsible for marking assignments

Students will send their completed exercises/assignments to the following email address: BME1478H@gmail.com

Password: rAdiohEad0699

**Comment by linamnt** [comment]
I removed the pass as this is a public repo...

**Comment by joelostblom** [comment]
@SaraMati @linamnt The comment's edit history is publicly visible if you click on `edited by...`, so you need to change the gmail password.

The above division of labor looks fins to me in principle, but I guess it is impossible to say if it is fair until we see what workload each task has. Just to clarify, the online support is not tied to the specific Wednesday class time, right?

Either slack or email is fine for me, slack has the advantage that the students can help each other and see the responses to questions posed by other students. Maybe Github issues in a dedicated support repo is a good solution? That has most of the advantages we want from slack and would not require any additional setup assuming that the students will create a GitHub account anyways, and it would get them a bit more familiar with GitHub. Private inquiries can be sent via email.

**Comment by joelostblom** [comment]
@SaraMati @linamnt I [deleted the content of the comment's history](https://help.github.com/en/github/building-a-strong-community/tracking-changes-in-a-comment#deleting-sensitive-information-from-a-comments-history), so there is no pressing need to change the password (although it is always good when it has been posted publicly, even just for a little bit, not sure how things like the way back machine indexes GitHub comments for example).

**Comment by SaraMati** [comment]
@joelostblom  
- do you receive emails at your utoronto email? 
- my bad regarding the password, but the repo was private in the beginning! I couldn't find the status of the repo now. we will change it on Monday. 
- Lina and I are going to have a meeting with Dr. Garton on Monday 6 pm time (EDT) probably, regarding the material. do you like to join in since it is about material? 
- I agree, I think a Github repo and using the issues is the best bet. another software may be overwhelming for students. BME1478H account can make another repo. we will discuss this on Monday

**Comment by joelostblom** [comment]
Yes I receive emails at the university address.

I'm not able to join at 6 pm unfortunately, I'm in Stockholm so that is
midnight over here =/ But most of my thoughts on the outline can be found
in the email I sent to you earlier so feel free to forward or discuss that
at the meeting as you see fit.

**Comment by linamnt** [comment]
@joelostblom 
> Just to clarify, the online support is not tied to the specific Wednesday class time, right?
No, it is not necessary as far as I can tell that you are available for support during the 10-12pm Wednesday class.

I think it's generally speaking when students post questions online (in whatever option we choose, GitHub, Slack etc.), you would primarily address them when you've allocated yourself the time to do so, presumably at least once per week. Sara, and I think some of the other TA's can assist with questions if/when the load becomes higher.

**Comment by joelostblom** [comment]
I saw Michal's email with the updated course outline. I don't know how much of this is still up for discussion, but since I can't make it to the meeting today, here are my comments on the new outline (in addition to the comments I emailed you previously).

## Suggestions for the outline

On a high level, I would group the lectures into three categories:

-   2-5: Basic Python concepts
-   6, 10-12: Data analysis
-   1, 7, 8: Supportive techniques

Based on that, I add some suggestions for the outline below trying to adhere to the current outline as much as possible (with the caveat that I don't know if you are teaching high performance computing to use it in the ML lessons or as a general tool).

1.  Intro computing, command line and how to access high performance computing resources.
2.  Python intro, concepts, development environments, first steps.
3.  Python datatypes, etc
4.  Python classes and oop.
5.  Troubleshooting and best practices / Debugging and version control (this could be made into a bigger point about sharing work openly, but that doesn't seem to be a focus of the course).
6.  Using Python instead of Excel, pandas and EDA concepts.
7.  Pandas + vis for EDA
8.  More on Vis (e.g. interactive, or an additional ML lecture before NNs, DTs or clustering?)
9.  Formal testing with statistics
10. Intro ML
11. NNs
12. NNs

## Lecture specific comments for the outline in the email

1.  I would also include a brief general introduction to computing. Even if these are engineering students, it is good to get everyone up to a general understanding. You can take what we use for the workshops I sent out previously if you want. Although I love vim myself, I would not teach it to an introductory class (maybe just how to get out of it if someone accidentally gets in there but I think that is covered in the git lesson). Likewise, I think it is more useful for student to use a text editor for search and replace instead of `sed`. IMO it is better to thoroughly explain topics like listing content, navigation, cp/mv/rm, option flags, how to get help, wildcards, briefly loops, some useful tools like `grep`, `head`, `top`/`htop`, (maybe `find` and `ssh` in preparation for working remotely without a graphical file namager). The purpose being to give a good basic understanding of how to work in the shell and hint at its more powerful feature to incentivize students to learn more on their own.

2.  I think one development environment, e.g. JupyterLab, should be focused, and others just mentioned or explained conceptually in how they differ and when they can be advantageous. Will git be explained here? If not, maybe move Github to the lecture 1 or 7? Depending on where `git` will be explained. Does this lecture include an intro to markdown and the underlying concepts of why it can be beneficial to use notebooks over IDEs?

3.  Will this be taught with a focus on data analysis or generally?

4.  Same as 3. Overall, I think having four lessons (lessons 2-5) on basic Python concepts is a bit much and it can probably be done in 2-3 lessons, which would give room for 1-2 additional data analysis specific lectures, such as pandas.

5.  I don't think an entire lecture on debugging is needed in an intro course in programming. Also, I previously wrote a paragraph on how to debug notebooks, pasting below so you can use it if you think it fits.

    > After you get an error in a notebook, you can use `%debug` in a new cell to enter post-mortem debugging with `pdb`. Here is a cheat sheet for [how to navigate the debugging prompt](https://appletree.or.kr/quick_reference_cards/Python/Python%20Debugger%20Cheatsheet.pdf) (most notably `u`/`d` to get to the level of the variables you want to check).

    > You can also set explicit breakpoints between lines of code with `breakpoint()` (standard library, no import needed). Now if only they fixed support for [tab-completion and history](https://github.com/jupyterlab/jupyterlab/issues/7006...)...

    > It might not be as sophisticated as a full blown graphical debugger, but especially the post-mortem can help compared to print statements. There are also traceback reformatters that automatically prints variables when an error is raised, such as [stackprinter](https://github.com/cknd/stackprinter), but I haven't found it as helpful as `%debug`.

    > [Some more tips and general discussion on debugging notebooks](https://stackoverflow.com/questions/32409629/what-is-the-right-way-to-debug-in-ipython-notebook).

6.  There will probably be enough material to just cover Python here and not other resources also, unless there is some specific novice friendly way to visualization method that you specifically want to teach of course.

7.  Is this where git will be taught? Are these best practices for Python coding or more general support tools around the coding?

8.  To me, this is something novices does not need to do in detail and this can be grouped with another lecture, such as debugging or shell.

9.  9-12 Looks good that there is now more lectures on ML. I think lecture 10 will be critical to link statistics to ML, explain the general concepts of ML, and motivate why we will be teaching NNs in the following two lectures instead of easier to understand simpler methods. I would pick the same high level framework for both 11 and 12, maybe keras or pytorch.
